### PR TITLE
Prevent crash with `WandbCallback` with third parties

### PR DESCRIPTION
**Pull request by tomaarsen**
# What does this PR do?

This PR fixes a dodgy if-statement introduced in #30135 that attempts to check whether a model has a `num_parameters` method. Currently, if
```python
(is_torch_available() and isinstance(model, torch.nn.Module))
```
is True, then the model does not necessarily have a `num_parameters` method. Instead, I've replaced it with an "Easier to ask for forgiveness than permission" principle which should be much safer.

This is currently breaking Sentence Transformers training with W&B.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@amyeroberts as she reviewed #30135
@muellerz 

- Tom Aarsen

### CODE REVIEW

File: src/transformers/integrations/integration_utils.py
Comment (amyeroberts) [inline src/transformers/integrations/integration_utils.py:792]: Could we add a logger.info here (feel free to change the message)? By default in transformers info level isn't logged, but might be useful if someone goes into debug mode 

```suggestion
                logger.info("Unable to get num_parameters from model")
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

File: src/transformers/integrations/integration_utils.py
Comment (tomaarsen) [reply src/transformers/integrations/integration_utils.py]: Good call, resolved via a6d09f523bf9241480e9d86f41c2a0ad4db880d2

**Comment by amyeroberts** [comment]
@tomaarsen Let me know if you need me to merge this in once CI is all green

**Comment by tomaarsen** [comment]
@amyeroberts I forget what the current merging protocol is for `transformers`. Do only the core maintainers merge? 
I'm good with either way :) 

- Tom Aarsen

**Comment by amyeroberts** [comment]
@tomaarsen Once a core maintainer has approved, anyone is free to merge, I just forget who has permission to

**Comment by tomaarsen** [comment]
I see! Makes sense :) I'll merge it.

- Tom Aarsen

**Pull request merged**
